### PR TITLE
[inductor] reusing autotuning sub-processes

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96729
* #97215

The major cost of doing autotuning in sub process is process creating and initialization. Previously we do that for each benchmark task. This PR reuse a child process as long as it has not crashed yet. This improves compiling time a lot. It's still a bit slower than single process tuning though. Here are the comparison between single process tuning and multi-process tuning:
- if a benchmark task will crash the process, then single process tuning is a no-go
- if a benchmark task works fine, then tuning in child process will be slower. We will try to leveraging multi-GPU to further speed this up.

TLDR for the compilation time: we reduce the 11x slowdown to 1.5x. We'll try to further improve that.

Here are the compilation time comparison:

Single process autotuning:
```
AUTOTUNE ref_mm_plus_mm(2048x64, 64x1536, 2048x64, 64x1536)
  triton_mm_plus_mm_0 0.0276s 100.0%
  triton_mm_plus_mm_6 0.0287s 96.4%
  triton_mm_plus_mm_5 0.0307s 90.0%
  triton_mm_plus_mm_1 0.0317s 87.1%
  triton_mm_plus_mm_7 0.0379s 73.0%
  ref_mm_plus_mm 0.0389s 71.1%
  triton_mm_plus_mm_2 0.0399s 69.2%
  triton_mm_plus_mm_3 0.0410s 67.5%
  triton_mm_plus_mm_4 0.0410s 67.5%
SingleProcess AUTOTUNE takes 9.04686689376831 seconds
```

Naive multi process tuning (not reuse child process): 11x slower than single process autotuning

```
AUTOTUNE ref_mm_plus_mm(2048x64, 64x1536, 2048x64, 64x1536)
  triton_mm_plus_mm_0 0.0287s 100.0%
  triton_mm_plus_mm_6 0.0287s 100.0%
  triton_mm_plus_mm_1 0.0317s 90.3%
  triton_mm_plus_mm_5 0.0317s 90.3%
  triton_mm_plus_mm_7 0.0379s 75.7%
  ref_mm_plus_mm 0.0389s 73.7%
  triton_mm_plus_mm_2 0.0399s 71.8%
  triton_mm_plus_mm_3 0.0399s 71.8%
  triton_mm_plus_mm_4 0.0420s 68.3%
SubProcess AUTOTUNE takes 101.22216320037842 seconds
```

Multi process tuning reusing child process (this PR): 1.5x slower than single process autotuning
```
AUTOTUNE ref_mm_plus_mm(2048x64, 64x1536, 2048x64, 64x1536)
  triton_mm_plus_mm_0 0.0276s 100.0%
  triton_mm_plus_mm_6 0.0287s 96.4%
  triton_mm_plus_mm_5 0.0307s 90.0%
  triton_mm_plus_mm_1 0.0317s 87.1%
  triton_mm_plus_mm_7 0.0379s 73.0%
  ref_mm_plus_mm 0.0389s 71.1%
  triton_mm_plus_mm_2 0.0399s 69.2%
  triton_mm_plus_mm_3 0.0410s 67.5%
  triton_mm_plus_mm_4 0.0410s 67.5%
SubProcess AUTOTUNE takes 13.752070665359497 seconds
```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Differential Revision: [D44050165](https://our.internmc.facebook.com/intern/diff/D44050165)

### CODE REVIEW

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44050165).

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44050165).

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44050165).

**Comment by shunting314** [comment]
I ran cmf_10x overarch with subprocess autotuning:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 TORCHINDUCTOR_AUTOTUNE_IN_SUBPROC=1 time buck run @mode/opt @mode/inplace //pytorch/benchmark:run -- cmf_10x -d cuda -t train --bs 1024 --torchdynamo inductor --part over
```

Now it talks around 1hour
```
WARNING:2023-03-14 23:09:02 3774893:3803149
```

There are 2 things to note:
1. I force not looking into the cache by return False directly in `torch._inductor.codecache.PersistentCache.lookup.<locals>.check_cache` return False . I'm not sure if that makes the reported number conservative. Let me know if there are better ways to disable cache for testing
2. Even though in my unit test, in shows subprocess autotuning is still around 1.5x slower, I found it may just because we need warmup the child process more. How long do does max_autotuning usually take for cmf_10x overarch if there was no crash? cc @bertmaher

**Comment by shunting314** [comment]
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44050165).

**Comment by shunting314** [comment]
resend PR
